### DISCUSSION TITLE
[v4] extending querybuilder, returning promise

### DISCUSSION THREAD

**Discussion by maxgalbu** [adonisjs/core#2324] [Help] [answered]
with the following example:

```
const slugify = require("./slugify");

class SlugifyTrait {
  register (Model, options) {
    Model.queryMacro('whereSlug', async function (value) {
      value = await slugify(value);
      this.where('slug', value)
      return this;
    })
  }
}

module.exports = SlugifyTrait;
```

I cannot do something like:
```
const user = await User.query().whereSlug("slug", name).first();
```

... of course, now whereSlug returns a promise, the error thrown is "Make sure to call fetch to execute the query" from proxyHandler.

Neither can i do, after commenting that line in proxyHandler:

```
let query = await User.query().whereSlug("slug", name);
const user = query.first();
```

...because query then it's [] 👎 

Is it not meant to do that? Is there a workaround or alternative solution? 🤕

**Comment by thetutlage** [comment, marked as answer]
The query builder methods are synchronous and not async. So creating `whereSlug` async method will simply not work. Either you have to move it the hook, or make the `whereSlug` synchronous

**Reply by maxgalbu** [reply]
For posterity, I've actually resorted to a very low level solution, that is creating a subclass of the existing knex driver and modifying the _query function - since it returns a promise, i can do a lot of the async stuff there. 

(obviously the example in the first post was a VEEERY BIG simplification of what i needed :) )
